src/Microsoft.OData.Core/UriParser/Binders/InBinder.cs

```diff
@@ -139,72 +132,144 @@ private CollectionNode GetCollectionOperandFromToken(QueryToken queryToken, IEdm
             return operand;
         }
 
-        private static string NormalizeCollectionItems(string bracketLiteralText, NormalizeFunction normalizeFunc)
+        private static string NormalizeStringCollectionItems(string literalText)
         {
-            string[] items = bracketLiteralText.Substring(1, bracketLiteralText.Length - 2).Split(',')
-                .Select(s => s.Trim()).ToArray();
+            // a comma-separated list of primitive values, enclosed in parentheses, or a single expression that resolves to a collection
+            // However, for String collection, we should process:
+            // 1) comma could be part of the string value
+            // 2) single quote could not be part of string value
+            // 3) double quote could be part of string value, double quote also could be the starting and ending character.
 
-            StringBuilder builder = new StringBuilder();
-            for (int i = 0; i < items.Length; i++)
+            // remove the "[" and "]"
+            string normalizedText = literalText.Substring(1, literalText.Length - 2).Trim();
+            int length = normalizedText.Length;
+            bool itemStarted = false;
+            int itemIndex = 0;
+            char[] updated = new char[length + 2];
+            updated[0] = '[';
+            char startingChar = normalizedText[0];
+            int j = 1;
+            for (int i = 0; i < length; i++)
             {
-                string convertedItem = normalizeFunc(items[i]);
-                if (i != items.Length - 1)
+                char character = normalizedText[i];
+                if (character == '"')

```

Comment:
I feel this is overly complex. I suspected the fix should have been simpler - 
1. We should not loop over the string again and again. You loop once to find item start and end index, then iterate over with the SubString method. My assumption is that SubString itself is O(n).

2. The logic should have been simple, create a string builder, if starting character is ' or ", keep on appending to sb until we encounter the same closing quote, unless it is double escaped. You can check by looking ahead one character. Once that an items end, a comma should follow if it is not the end of normalized text.

Reply:
About point 2.: as far as I can tell OData allows whitespace between the string terminator and the comma, so "... in ('alice'  ,  'bob')" should be legal. Is this correct?

Reply:
1) To loop over the string again and again is the original goal. In the implementation, there are only one loop. Which one do you think we loop it twice. For the SubString, you can check it only happens when there's a mismatch, and throw exception.  In order to give the customer which item is wrong, i have to use "substring" to retrieve the "sub-string".

2) For the logic, i think my logic is same as your suggestion, but I use a string[] and not a stringbuilder. I don't want to use stringbuilder because stringbuilder is not fast than string[]. Besides, the key problem:

1) double quote can be wrapper of the string, or can be part of the string value.
 for example:   "(\" afbc'efg\")" and "('   afbc\"efg')"  are valid in "In" operator
   but, "(\" afbc'efg')', "(' afbc'efg\")" are not valid.

2) single quote can be wrapper of the string, or can be part of the string value. But, if single quote is par of the string, it should be "encoded" as double single quote. 

3) Comma can be separator of items, or can be part of the string value.  
        



---
In reply to: [367803756](https://github.com/OData/odata.net/pull/1635#discussion_r367803756) [](ancestors = 367803756)

Reply:
@Kenneth Aas Hansen, the white space is allowed between string terminator. that's my thought. So, in my implementation, i remove the whitespace at Line 233.

I want to note that:

1) If you double commas, for example:  "... in ('alice'  , ,, 'bob'), i will let it go and let the low level process to identify the error or ignore the items between the commas.

2) for all other character between terminator, i throw exception, at Line 243.




---
In reply to: [367809728](https://github.com/OData/odata.net/pull/1635#discussion_r367809728) [](ancestors = 367809728)